Automattic/wp-calypso
# Checkout: Show introductory offer details in discount list

## Proposed Changes

This PR updates the sidebar discount list to include details for Introductory offers. Instead of just reading "Introductory offer", it will now read something like "Discount for first year".

Here's a Jetpack introductory offer:

Before             |  After
:-------------------------:|:-------------------------:
<img width="335" alt="Screenshot 2024-02-05 at 12 10 35 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/c2b91172-d61e-48aa-bb6f-658c3b19d350"> | <img width="332" alt="Screenshot 2024-02-05 at 12 10 56 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/977b480d-2642-429a-9777-e24211479aad">

Here's a WPCom introductory offer (note that we also merge Sale coupons, which we should have done originally in https://github.com/Automattic/wp-calypso/pull/86371):

Before             |  After
:-------------------------:|:-------------------------:
<img width="321" alt="Screenshot 2024-02-05 at 12 20 45 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/333b4be9-e452-4527-be11-cc472a203d91"> | <img width="314" alt="Screenshot 2024-02-05 at 1 44 55 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/4e023af7-7311-4ac4-bb15-b41cdfd6fc59">


> [!NOTE]
> This preserves the multi-year discount for Jetpack biennial and hiding introductory offers for Jetpack social added by https://github.com/Automattic/wp-calypso/pull/86353. They've been made modular so we could change or extend those features to other products more easily in the future (eg: showing multi-year discounts for WPCom plans), although doing so is outside the scope of this PR.

> [!NOTE]
> Jetpack introductory offers showed an asterisk next to the "Introductory offer" item and then a notice at the bottom of the list that explains the offer is valid for the first term only. I've retained this behavior but fixed it to only display if the introductory offer is indeed for the first term. There are some issues with this notice (https://github.com/Automattic/wp-calypso/issues/87194) but they are outside the scope of this PR.

Here's Jetpack with a multi-year discount:

Before             |  After
:-------------------------:|:-------------------------:
<img width="318" alt="Screenshot 2024-02-05 at 12 12 26 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/eef2176a-b9d4-486e-846b-98a31dab3d9c"> | <img width="314" alt="Screenshot 2024-02-05 at 12 12 36 PM" src="https://github.com/Automattic/wp-calypso/assets/2036909/ea9cd582-ffef-43ac-87c1-0da84599302e">

Fixes https://github.com/Automattic/wp-calypso/issues/87108

## Testing Instructions

- Add products to your cart with introductory offers (eg: most Jetpack products, or Titan mail).
- Verify that the subtotals and totals appear the same.
- Verify that the sidebar discount list shows details about each offer in a way that makes sense.
- Verify that the price in the sidebar for those offers is correct.
